src/placeos-driver/settings/introspect.cr

```diff
@@ -0,0 +1,114 @@
+            {% for key, ivar in properties %}
+              {{key}}: PlaceOS.introspect({{ivar.resolve.name}}),
+            {% end %}
+          }, required: [
+            {% for key, ivar in properties %}
+              {% if !ivar.nilable? %}
+                {{key.stringify}},
+              {% end %}
+            {% end %}
+          ] of String}
+        {% end %}
+      {% end %}
+    end
+  end
+
+  module ::JSON::Serializable
+    macro included
+      extend ::PlaceOS::Introspect
+    end
+  end
+
+  macro introspect(klass)
+    {% arg_name = klass.stringify %}
+    {% if !arg_name.starts_with?("Union") && arg_name.includes?("|") %}
+      PlaceOS.introspect(Union({{klass}}))
+    {% else %}
+      {% klass = klass.resolve %}
+      {% klass_name = klass.name(generic_args: false) %}
+
+      {% if klass <= Array %}
+        {% if klass.type_vars.size == 1 %}
+          has_items = PlaceOS.introspect {{klass.type_vars[0]}}
+        {% else %}
+          has_items = {} of String => String
+        {% end %}
+        if has_items.empty?
+          {type: "array"}
+        else
+          {type: "array", items: has_items}
+        end
+      {% elsif klass.union? %}
+        { anyOf: [
+          {% for type in klass.union_types %}
+            PlaceOS.introspect({{type}}),
+          {% end %}
+        ]}
+      {% elsif klass_name.starts_with? "Tuple(" %}
+        has_items = [
+          {% for generic in klass.type_vars %}
+            PlaceOS.introspect({{generic}}),
+          {% end %}
+        ]
+        {type: "array", items: has_items}
+      {% elsif klass_name.starts_with? "NamedTuple(" %}
+        {type: "object",  properties: {
+          {% for key in klass.keys %}
+            {{key.id}}: PlaceOS.introspect({{klass[key].resolve.name}}),
+          {% end %}
+        }, required: [
+          {% for key in klass.keys %}
+            {% if !klass[key].resolve.nilable? %}
+              {{key.id.stringify}},
+            {% end %}
+          {% end %}
+        ] of String}
+      {% elsif klass < Enum %}
+        {type: "string",  enum: {{klass.constants.map(&.stringify)}} }
+      {% elsif klass <= String %}
+        { type: "string" }
+      {% elsif klass <= Bool %}
+        { type: "boolean" }
+      {% elsif klass <= Int %}
+        { type: "integer" }
+      {% elsif klass <= Float %}
+        { type: "number" }
+      {% elsif klass <= Nil %}
+        { type: "null" }
+      {% elsif klass <= Hash %}
+        {% if klass.type_vars.size == 2 %}
+          { type: "object", additionalProperties: PlaceOS.introspect({{klass.type_vars[1]}}) }
+        {% else %}
+          # As inheritance might include the type_vars it's hard to work them out
+          { type: "object" }
+        {% end %}
+      {% elsif klass.ancestors.includes? JSON::Serializable %}
+        {{klass}}.__generate_json_schema__
+      {% else %}
+        # anything will validate (JSON::Any)
+        {} of String => String
```

Comment:
I think we should also provide helper methods to build a custom json schema for objects that we know will be serialised from this very generic branch, or even in cases where we want to override the default serialisation. so for example

```
macro write_schema(schema)
  # do some stuff here  
end

class SuperArray < Array(Int32)
  write_schema({type: "array", items: {type: "string"}})
end
```

Reply:
added some basic support for custom schemas where the class is custom